### ISSUE TITLE
Reactor ignoring Controlnet

### ISSUE THREAD

**Issue by Ethan-Ming** [Gourieff/sd-webui-reactor#287]
### First, confirm

- [X] I have read the [instruction](https://github.com/Gourieff/sd-webui-reactor/blob/main/README.md) carefully
- [X] I have searched the existing issues
- [X] I have updated the extension to the latest version

### What happened?

the source image has eyes closed, using reactor only will cause the generated image to have eyes opened AND expression influenced by reactor source image instead of original image.

usually, this is fixable by using ControlNet, but no matter what ControlNet model i try (mainly MediaPipe_face and openpose_face_only). the reactor seems keeps ignoring ControlNet input

![Ca142354245pture](https://github.com/Gourieff/sd-webui-reactor/assets/36059080/c4ab6b61-817f-414b-a9c8-28e305b43e80)

![Cap2353245ture](https://github.com/Gourieff/sd-webui-reactor/assets/36059080/7243e1e9-33ae-457f-b4ee-ed289d9c683d)

### Steps to reproduce the problem

1. Go to img2img
2. set controlnet, use mediapipe_face model. as in the image
3. set reactor as recommended

4. if doing it correctly, you should see face swapped but Reactor is ignoring the facial expression 

### Sysinfo

WIN10, CHROME browser

version: [v1.6.0](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/5ef669de080814067961f28357256e8fe27544f4)  •  python: 3.10.10  •  torch: 2.0.0+cu118  •  xformers: N/A  •  gradio: 3.41.2  •  checkpoint: [cf65f48ecf](https://google.com/search?q=cf65f48ecfd2fb6b7b8a83eb3f99448d6e85b1ef4261c2f9ecc743d0f72decf0)

### Relevant console log

```Shell
15:12:55 - ReActor - STATUS - Running v0.6.0-a4
[AddNet] Updating model hashes...
0it [00:00, ?it/s]
[AddNet] Updating model hashes...
0it [00:00, ?it/s]
ControlNet preprocessor location: E:\AUTOMATIC1111\stable-diffusion-webui\extensions-builtin\sd-webui-controlnet\annotator\downloads
2023-12-28 15:13:25,008 - ControlNet - INFO - ControlNet v1.1.423
2023-12-28 15:13:25,449 - ControlNet - INFO - ControlNet v1.1.423
Loading weights [cf65f48ecf] from E:\AUTOMATIC1111\stable-diffusion-webui\models\Stable-diffusion\SD2.1_replicantV30_v30.safetensors
Creating model from config: E:\AUTOMATIC1111\stable-diffusion-webui\repositories\stable-diffusion-stability-ai\configs\stable-diffusion\v2-inference-v.yaml
Applying attention optimization: sdp... done.
Model loaded in 10.2s (load weights from disk: 0.5s, find config: 5.3s, create model: 0.2s, apply weights to model: 3.1s, apply float(): 0.1s, move model to device: 0.1s, load textual inversion embeddings: 0.6s, calculate empty prompt: 0.3s).
Running on local URL:  http://0.0.0.0:7860
------
[00:10<00:00, 10.25s/it]
2023-12-28 15:50:03,721 - ControlNet - INFO - unit_separate = False, style_align = False█| 1/1 [00:10<00:00, 10.25s/it]
2023-12-28 15:50:04,046 - ControlNet - INFO - Loading model: control_v2p_sd21_mediapipe_face [6e0c43ab]
2023-12-28 15:50:04,114 - ControlNet - INFO - Loaded state_dict from [E:\AUTOMATIC1111\stable-diffusion-webui\extensions-builtin\sd-webui-controlnet\models\control_v2p_sd21_mediapipe_face.safetensors]
2023-12-28 15:50:04,114 - ControlNet - INFO - controlnet_default_config
2023-12-28 15:50:09,452 - ControlNet - INFO - ControlNet model control_v2p_sd21_mediapipe_face [6e0c43ab] loaded.
2023-12-28 15:50:09,563 - ControlNet - INFO - Loading preprocessor: mediapipe_face
2023-12-28 15:50:09,564 - ControlNet - INFO - preprocessor resolution = 1062
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
2023-12-28 15:50:13,117 - ControlNet - INFO - ControlNet Hooked - Time = 9.401198387145996
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.90s/it]
15:50:23 - ReActor - STATUS - Working: source face index [0], target face index [0]              | 0/1 [00:00<?, ?it/s]
15:50:23 - ReActor - STATUS - Analyzing Source Image...
15:50:25 - ReActor - STATUS - Analyzing Target Image...
15:50:26 - ReActor - STATUS - Detecting Source Face, Index = 0
15:50:26 - ReActor - STATUS - Detected: -37- y.o. Female
15:50:26 - ReActor - STATUS - Detecting Target Face, Index = 0
15:50:26 - ReActor - STATUS - Detected: -29- y.o. Female
15:50:26 - ReActor - STATUS - Swapping Source into Target
15:50:27 - ReActor - STATUS - Correcting Face Mask
15:50:27 - ReActor - STATUS - Restoring the face with CodeFormer
Total progress: 100%|████████████████████████████████████████████████████████████████████| 1/1 [00:12<00:00, 12.64s/it]
Total progress: 100%|████████████████████████████████████████████████████████████████████| 1/1 [00:12<00:00, 12.64s/it]
```

**Comment by Gourieff** [comment]
ReActor doesn't ignores ControlNet
There's no feature yet to detect target's gaze
In addition, the face restoration model is also used (CodeFormer, GFPGAN) which can cause "eyes-opening" as well

**Comment by Ethan-Ming** [comment]
gazing control is welcome, but i'm aware of the implementation aren't there yet so I intended to discuss something else, 

the expressions, eyelids, and directions of the face were detected by controlnet as by now

so is possible to make the rector respect controlnet input more?

I've just set Control Weight to 2 and ran the test again, seems did very little effect if not none

BTW, just tried turning the face restoration model off with few more images,the "eyes-opening" seems worsened. for me leaving it on yield better results

**Comment by Ethan-Ming** [comment]
> ReActor doesn't ignores ControlNet There's no feature yet to detect target's gaze In addition, the face restoration model is also used (CodeFormer, GFPGAN) which can cause "eyes-opening" as well

 gazing control is welcome, but i'm aware of the implementation aren't there yet so I intended to discuss something else, 

the expressions, eyelids, and directions of the face were detected by controlnet as by now

so is possible to make the rector respect controlnet input more?

I've just set Control Weight to 2 and ran the test again, seems did very little effect if not none



BTW, just tried turning the face restoration model off with few more images,the "eyes-opening" seems worsened. for me leaving it on yield better results

**Comment by listvin** [comment]
To my understanding ReActor has nothing to do with ControlNet or image generation at all.
E.g. in txt2img mode, using controlnet and reactor does following:

1. İmage is generated — ControlNet hooks to stable diffusion directly and they together generate image.
<sd and cnet models can even be offloaded here from vram>
2. Output from step 1 goes into ReActor, it does it job, replacing image

Step 1 can be img2img or whatever
There can be now step 1 at all, if you use ReActor from Extras tab, just with any input image, not necessarily ai generated.

I believe hooking ControlNet into face swap model is some ai engineering going beyond the exact extension

**Comment by Gourieff** [comment]
@listvin perfect explanation 👍

**Comment by Ethan-Ming** [comment]
@listvin if i'm reading it correctly, the "issue" is due to ControlNet generating image **BEFORE** the reactor, so for do what we wanted,  ControlNet needs to post-processing the image **AFTER** reActor generated its image-_(i hope i said that right)_-therefor i assume this can be done with some clever noding skills in comfy ui?

**Comment by Gourieff** [comment]
ControlNet is for processing the image (as the part of the image generation process), ReActor is for postprocessing, therefore ControlNet cannot be run after ReActor, even if we build the pipeline in ComfyUI
Precisely speaking, technically it may be launched like that in ComfyUI, we can swap smth with ReActor than send the result to ControlNet... But what's the point? The face will be rebuilt with a checkpoint model and we will need to run ReActor again
Actually, we can create a reverse inpaint mask to "freeze" the face and to try to redraw everything except the face... But it will hardly solve the "eyes-opening" issue
